src/libraries/System.Net.Http/ref/System.Net.Http.cs

```diff
@@ -901,3 +905,19 @@ public partial class WarningHeaderValue : System.ICloneable
         public static bool TryParse([System.Diagnostics.CodeAnalysis.NotNullWhenAttribute(true)] string? input, [System.Diagnostics.CodeAnalysis.NotNullWhenAttribute(true)] out System.Net.Http.Headers.WarningHeaderValue? parsedValue) { throw null; }
     }
 }
+namespace System.Net.Http.Metrics
+{
+    public class HttpMetricsEnrichmentContext
+    {
+        public HttpRequestMessage Request { get { throw null; } }
+
+        public HttpResponseMessage? Response { get { throw null; } }
+
+        public Exception? Exception { get { throw null; } }
+        public ICollection<KeyValuePair<string, object?>> CustomTags { get { throw null; } }
+    }
+    public static class MetricsHttpRequestOptionsExtensions
+    {
+        public static void AddMetricsEnrichmentCallback(this HttpRequestOptions options, Action<HttpMetricsEnrichmentContext> callback) { throw null; }
```

Comment:
This design means the actual callbacks aren't exposed, which means a handler other than one that lives in System.Net.Http can't fully implement this feature.  That might very well be the right tradeoff, but we should know that's what we're doing here.

Reply:
The API is defined against `HttpRequestOptions` where storing handler-specific things should be ok IMO.
If other handlers would want to hook into this in the future, we can define another extension method to run the callbacks and collect the tags, or expose `MetricsHandler` publicly.